lib/spack/spack/package.py

```diff
@@ -1420,6 +1524,43 @@ def rpath_args(self):
         return " ".join("-Wl,-rpath,%s" % p for p in self.rpath)
 
 
+class Package(PackageBase):
+    phases = ['install', 'log']
+    # This will be used as a registration decorator in user
+    # packages, if need be
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class AutotoolsPackage(PackageBase):
+    phases = ['autoreconf', 'configure', 'build', 'install', 'log']
+
+    def autoreconf(self, spec, prefix):
+        """Not needed usually, configure should be already there"""
+        pass
+
+    @PackageBase.sanity_check('autoreconf')
+    def is_configure_or_die(self):
+        if not os.path.exists('configure'):
+            raise RuntimeError('configure script not found in {0}'.format(os.getcwd()))
+
+    def configure_args(self):
+        return list()
```

Comment:
@adamjstewart @davydden 

> configure_args definitely needs spec passed to it since that dictates dependency prefixes and which flags to use. Might as well pass prefix as well to make things uniform

I don't have strong opinions on this : if it makes sense to most packagers then changing the arguments will be fine with me too. Anyhow I always found this : 
```python
def install(self, spec, prefix):
    pass
```
a bit useless. I would have preferred this to be generated instead: 
```python
def install(self):
    spec = self.spec
    prefix = self.prefix
    ...
```
as I never found a context in which `spec != self.spec` and `prefix != self.prefix`.